### ISSUE TITLE
DropwizardExecutorProvider not shareable but it is shared 

### ISSUE THREAD

**Issue by Kinchkun** [dropwizard/dropwizard#2218]
Hello,

I have the following behavior in dw 1.1.4 and 1.2.0. I create a JerseyClient and bind it to hk2 via:

```
final Client client = new JerseyClientBuilder(environment)
    .using(configuration.getJerseyClientConfiguration())
    .build(smartMobileApplication.getName());
bind(client).to(Client.class);
```

Eventually I perform some async calls:

```
    client.target("http://example")
                .path("/example")
                .register(HttpAuthenticationFeature.basic(
                        configurationCredentials.getLogin(),
                        configurationCredentials.getPass()
                ))
                .request()
                .async()
                .post(Entity.entity(entity.getPayload(), MediaType.APPLICATION_JSON), callback);
    }
```

After some calls the Executor service is shutdown by hk2 leading to a `RejectedExecutionException`.

If found that removing the line
```
                .register(HttpAuthenticationFeature.basic(
                        configurationCredentials.getLogin(),
                        configurationCredentials.getPass()
                ))
```

from the code above fixes this issue.

I assume that the "register" call will alter the underlying `ClientConfig`, which leads to a copy from the jersey runtime (i believe they have a copy on write). The jersey runtime will cleanup these copys and the `DropwizardExecutorProvider` will shutdown. In the javadoc for the `DropwizardExecutorProvider` it is stated that the `DropwizardExecutorProvider` should not be shared. However, if think it is pretty intransparent that the `DropwizardExecutorProvider` is shared just by adding a `register` call to the `WebTarget`.

I am pretty new to dropwizard and java in general. Is this a bug or expected behaviour? Or do I misuse the client api?

Kind regards
Timo

**Comment by arteam** [comment]
Thanks for reporting the bug! I can confirm it. The reason behind is a weird behavior of HK, which is looking for dead references in the finalizer and then disposes them automatically (poor man's leak detector). This issue is similar to #1160. 

I think we should partially revert #1676 and remove the part about disposing the executor and let the Dropwizard environment manage it.

**Comment by arteam** [comment]
As a workaround you can provide a non-disposable executor during building the client:
```java
Client client = new JerseyClientBuilder(environment)
    .using(environment.lifecycle().executorService("async-executor")
    .minThreads(4)
    .maxThreads(4)
    .build())
.using(jc)
.build("jersey-test");
```

**Comment by Kinchkun** [comment]
Thanks for your fast response and workaround. I can confirm that the workaround works :)

**Comment by arteam** [comment]
Closed  via #2221
